nexus/src/db/datastore/resource_usage.rs

```diff
@@ -0,0 +1,221 @@
+            .iter()
+            .map(|usage| {
+                Sample::new(
+                    &CollectionTarget { id: usage.id },
+                    &CpusProvisioned { cpus: usage.cpus_provisioned },
+                )
+            })
+            .chain(usages.iter().map(|usage| {
+                Sample::new(
+                    &CollectionTarget { id: usage.id },
+                    &RamProvisioned { bytes: usage.ram_provisioned },
+                )
+            }))
+            .collect::<Vec<_>>();
+
+        self.append(new_samples);
+    }
+
+    fn append(&self, mut new_samples: Vec<Sample>) {
+        let mut pending_samples = self.samples.lock().unwrap();
+        pending_samples.append(&mut new_samples);
+    }
+}
+
+impl oximeter::Producer for Producer {
+    fn produce(
+        &mut self,
+    ) -> Result<Box<dyn Iterator<Item = Sample> + 'static>, MetricsError> {
+        let samples =
+            std::mem::replace(&mut *self.samples.lock().unwrap(), vec![]);
+        Ok(Box::new(samples.into_iter()))
+    }
+}
+
+impl DataStore {
+    /// Create a resource_usage
+    pub async fn resource_usage_create(
+        &self,
+        opctx: &OpContext,
+        resource_usage: ResourceUsage,
+    ) -> Result<Vec<ResourceUsage>, Error> {
+        let pool = self.pool_authorized(opctx).await?;
+        self.resource_usage_create_on_connection(pool, resource_usage).await
+    }
+
+    pub(crate) async fn resource_usage_create_on_connection<ConnErr>(
+        &self,
+        conn: &(impl async_bb8_diesel::AsyncConnection<DbConnection, ConnErr>
+              + Sync),
+        resource_usage: ResourceUsage,
+    ) -> Result<Vec<ResourceUsage>, Error>
+    where
+        ConnErr: From<diesel::result::Error> + Send + 'static,
+        PoolError: From<ConnErr>,
+    {
+        use db::schema::resource_usage::dsl;
+
+        let usages: Vec<ResourceUsage> =
+            diesel::insert_into(dsl::resource_usage)
+                .values(resource_usage)
+                .on_conflict_do_nothing()
+                .get_results_async(conn)
+                .await
+                .map_err(|e| {
+                    public_error_from_diesel_pool(
+                        PoolError::from(e),
+                        ErrorHandler::Server,
+                    )
+                })?;
+        self.resource_usage_producer.append_disk_metrics(&usages);
+        self.resource_usage_producer.append_cpu_metrics(&usages);
+        Ok(usages)
+    }
+
+    pub async fn resource_usage_get(
+        &self,
+        opctx: &OpContext,
+        id: Uuid,
+    ) -> Result<ResourceUsage, Error> {
+        use db::schema::resource_usage::dsl;
+
+        let resource_usage = dsl::resource_usage
+            .find(id)
+            .select(ResourceUsage::as_select())
+            .get_result_async(self.pool_authorized(opctx).await?)
+            .await
+            .map_err(|e| {
+                public_error_from_diesel_pool(e, ErrorHandler::Server)
```

Comment:
Should this produce NotFound?  (or: why do we (or the caller) know that the record necessarily exists?)

Reply:
Although all the current usages are from tests, I've added this in f05296e79298bb205cdd440a0730b104eed4b032 - just in case a new caller comes along.